### ISSUE TITLE
Convert all docstrings to numpy style

### ISSUE THREAD

**Issue by vyasr** [glotzerlab/signac#315]
As part of our overall docs overhaul (see glotzerlab/signac-docs#64), we want to convert our docstrings to numpy style (as decided in glotzerlab/signac-docs#74). The best automated tool I'm familiar with for this task is [pyment](https://github.com/dadadel/pyment). In addition to converting docstrings, it will also generate docstrings for functions, classes, etc that are missing docstrings entirely. However, the conversion will require significant manual review to ensure that all docstrings are converted correctly.

**Comment by vyasr** [comment]
@ac-optimus could you comment on this issue so that I can assign you?

**Comment by ac-optimus** [comment]
hi, can I work on this issue?

**Comment by mikemhenry** [comment]
List of files that need to be fixed:

- [ ] `benchmark.py`
- [ ] `doc/conf.py`
- [ ] `setup.py`
- [ ] `signac/common/configobj/__init__.py`
- [ ] `signac/common/configobj/validate.py`
- [ ] `signac/common/configobj/_version.py`
- [ ] `signac/common/config.py`
- [ ] `signac/common/connection.py`
- [ ] `signac/common/crypt.py`
- [ ] `signac/common/host.py`
- [ ] `signac/common/__init__.py`
- [ ] `signac/common/validate.py`
- [ ] `signac/contrib/filesystems.py`
- [ ] `signac/contrib/import_export.py`
- [ ] `signac/contrib/indexing.py`
- [ ] `signac/contrib/__init__.py`
- [ ] `signac/contrib/migration/__init__.py`
- [ ] `signac/contrib/migration/v0_to_v1.py`
- [ ] `signac/contrib/mpipool.py`
- [ ] `signac/core/attrdict.py`
- [ ] `signac/core/dict_manager.py`
- [ ] `signac/core/h5store.py`
- [ ] `signac/core/__init__.py`
- [ ] `signac/core/json.py`
- [ ] `signac/core/synceddict.py`
- [ ] `signac/db/database.py`
- [ ] `signac/db/__init__.py`
- [ ] `signac/diff.py`
- [ ] `signac/__init__.py`
- [ ] `signac/__main__.py`
- [ ] `signac/sync.py`
- [ ] `signac/syncutil.py`
- [ ] `signac/testing.py`
- [ ] `signac/version.py`
- [ ] `signac/warnings.py`
- [ ] `.sync-zenodo-metadata.py`

**Comment by bdice** [comment]
@mikemhenry Fantastic! We can also ~cross out~ any files that are deprecated and add them to a "disallow" list for docstyle checks, in addition to our current "allow" list.

**Comment by mikemhenry** [comment]
I'm thinking any files that are not in `signac/` can be probably dropped? Also any files that "`THIS MODULE IS DEPRECATED!`"

**Comment by mikemhenry** [comment]
Okay I think I got all the files that we should ignore, if anyone wants to propose/suggest others just let me know!

**Comment by csadorf** [comment]
I don't think we have to adapt the doc-strings of vendored packages such as the configobj package.

**Comment by vyasr** [comment]
I agree, we should leave those as is.

**Comment by bdice** [comment]
I edited @mikemhenry's comment directly to fix the vendored list. Good call. We could split up the list into 4 PRs: `common`, `contrib`, `core`, and "other".

**Comment by mikemhenry** [comment]
Thanks @bdice do you mean split the list up for organization or do you mean that we have 4 PRs that cover all of those folders/files instead of a PR-per-file

**Comment by bdice** [comment]
@mikemhenry I was considering that as 4 "blocks" of work that could each be assigned to a specific person. The choice of one PR per "block" or one PR per file could be up to the assignee.

**Comment by mikemhenry** [comment]
Got it!

**Comment by mikemhenry** [comment]
Didn't mean to close  I will update the list with those headings!

**Comment by bdice** [comment]
I did the `core` submodule in PR #339. We'll need to render all the docs at the end and make sure it looks good before merging the `numpy_docs` branch -- there have been a few things that I've review/approved/written just from looking at the Python code without looking at how it is rendered. I think this will be easiest to do all at the end.

**Comment by bdice** [comment]
I finished up #333, which is waiting on review.

**Comment by bdice** [comment]
I created #343. After #343 is merged, remaining PRs should remove `core`, `common`, and `contrib` from the "skipped" directories in `match-dir` in `setup.cfg`.

**Comment by bdice** [comment]
I think we may be ready to merge `numpy_docs` to the main branch after #352. We'll need a final review of the file list above, but importantly, `pydocstyle` should pass on the whole code base (excluding deprecations) after #352 is merged. This will allow us to start enforcing docstring style in some fashion for other pull requests, particularly #336.

1. The `core` submodule is being heavily rewritten in #336, so I don't think it is a good use of developer hours to update the existing `core` docstrings to NumPy docstring style right now.

2. The `common` submodule is low priority: it has many deprecated features, its functionality is almost strictly for internal purposes, and it is _not in the public API documentation at all_ (there is one un-linked reference to a `Config` object).

3. The `contrib` submodule has already had significant rewriting and appears to be completely adherent to NumPy docstring style with the exception of deprecated modules.

**Comment by bdice** [comment]
@glotzerlab/signac-committers I've opened PR #357, which will finish up much of the work that has been done for this issue. We'll need to thoroughly review the discussion on this issue. I would advocate for us to review #357 as it is, merge it (unless critical problems are found), and push all follow-up work to new issues/PRs.

Copying from a comment @b-butler made on #354: there are a couple methods with ```# noqa: D103``` because I couldn't figure out what to write for their docstrings and wanted pydocstyle to pass without proper docstrings. https://github.com/glotzerlab/signac/blob/ee80d8b130525d6c831ac7d8e5eb689de7d7690c/signac/common/validate.py#L33-L41

Also related to what I was saying above: the `common` submodule has many candidates for public API deprecations, in my estimation.
